### PR TITLE
Add opentelemetry capability to the runtime

### PR DESCRIPTION
**Pull request by haixuanTao**
This commit adds the ability to use the `opentelemetry` feature flag at runtime by using the env variable `DORA_TELEMETRY`. Using environment variable reduce the need to do some argument drilling between our different process as well as making configurable in the YAML at no cost.

Usage:

```bash
docker run -d -p6831:6831/udp -p6832:6832/udp -p16686:16686 jaegertracing/all-in-one:latest
DORA_JAEGER_TRACING=172.17.0.1:6831 dora-daemon --run-dataflow dataflow.yml
```

If Jaeger variable is not set, we will default to stdout:

```bash
RUST_LOG=trace dora-daemon --run-dataflow dataflow.yml
```

Main change in this pull request:
- use tracing_opentelemetry to unify our tracing ecosystem and opentelemetry ecosystem. This enables me to remove locally declared tracer. 
- This also push our current tracing instrumentation into Jaeger when the env variable is activated. ( e.g. The tracing of dora-daemon, Dora-coordinator,...) 
- I have kept the possibility to fully deactivate opentelemetry ( through feature flag) if we want to make the application lighter. 

Not included in this PR and will be available in a future PR is making it easy within the user space to leverage the current opentelemetry context.

### CODE REVIEW

**Review by phil-opp — APPROVED** [review]
Looks good! We should fix the remaining import warnings though (by moving them to the #cfg block).

File: libraries/extensions/telemetry/tracing/src/lib.rs [hunk 1]
```diff
@@ -4,19 +4,53 @@
...
+                    .to_str()
+                    .wrap_err("Could not parse env variable: DORA_JAEGER_TRACING")?;
+                let tracer = crate::telemetry::init_jaeger_tracing(name, endpoint)
+                    .wrap_err("Could not instantiate tracing")?;
+                tracing_opentelemetry::layer()
+                    .with_tracer(tracer)
+                    .with_filter(filter.add_directive(LevelFilter::TRACE.into()))
+            }
+            None => {
+                let tracer = stdout::new_pipeline().install_simple();
```
Comment (phil-opp) [inline libraries/extensions/telemetry/tracing/src/lib.rs:39]: Is this `stdout` trace in a human-readable format?

Comment (haixuanTao) [reply libraries/extensions/telemetry/tracing/src/lib.rs]: So, I added the pretty print option in the `stdout::new_pipeline` from opentelemetry, which makes it better, 
From:
```bash
SpanData { span_context: SpanContext { trace_id: 0a6e102fea3d339013e23164c2394cb4, span_id: fb64a8354f00854f, trace_flags: TraceFlags(1), is_remote: false, trace_state: TraceState(None) }, parent_span_id: 0000000000000000, span_kind: Internal, name: "handle_connection_loop", start_time: SystemTime { tv_sec: 1679885600, tv_nsec: 94738878 }, end_time: SystemTime { tv_sec: 1679885600, tv_nsec: 739831257 }, attributes: EvictedHashMap { map: {Static("thread.id"): I64(2), Static("idle_ns"): I64(644847345), Static("queue_sizes"): String(Owned("{DataId(\"op/latency\"): 10}")), Static("code.namespace"): String(Static("dora_daemon::listener::tcp")), Static("code.filepath"): String(Static("binaries/daemon/src/listener/tcp.rs")), Static("busy_ns"): I64(243993), Static("code.lineno"): I64(44), Static("thread.name"): String(Owned("tokio-runtime-worker"))}, evict_list: [Static("idle_ns"), Static("busy_ns"), Static("queue_sizes"), Static("thread.name"), Static("thread.id"), Static("code.lineno"), Static("code.namespace"), Static("code.filepath")], max_len: 128, dropped_count: 0 }, events: EvictedQueue { queue: Some([Event { name: "channel disconnected: a264d253-4d9b-449f-ab44-d47343f42939/op_2", timestamp: SystemTime { tv_sec: 1679885600, tv_nsec: 739779000 }, attributes: [KeyValue { key: Static("level"), value: String(Static("DEBUG")) }, KeyValue { key: Static("target"), value: String(Static("dora_daemon::listener")) }, KeyValue { key: Static("code.filepath"), value: String(Static("binaries/daemon/src/listener/mod.rs")) }, KeyValue { key: Static("code.namespace"), value: String(Static("dora_daemon::listener")) }, KeyValue { key: Static("code.lineno"), value: I64(190) }], dropped_attributes_count: 0 }]), max_len: 128, dropped_count: 0 }, links: EvictedQueue { queue: None, max_len: 128, dropped_count: 0 }, status: Unset, resource: Resource { attrs: {Static("service.name"): String(Owned("unknown_service"))}, schema_url: None }, instrumentation_lib: InstrumentationLibrary { name: "opentelemetry", version: Some("0.18.0"), schema_url: None } }
```

To:
```bash
SpanData {
    span_context: SpanContext {
        trace_id: b42f990b8b761cbf4331b1b12393a168,
        span_id: 2c00b6ddc59e80ad,
        trace_flags: TraceFlags(
            1,
        ),
        is_remote: false,
        trace_state: TraceState(
            None,
        ),
    },
    parent_span_id: 0000000000000000,
    span_kind: Internal,
    name: "listener_loop",
    start_time: SystemTime {
        tv_sec: 1679885225,
        tv_nsec: 530300305,
    },
    end_time: SystemTime {
        tv_sec: 1679885226,
        tv_nsec: 974209562,
    },
    attributes: EvictedHashMap {
        map: {
            Static(
                "thread.id",
            ): I64(
                12,
            ),
            Static(
                "code.lineno",
            ): I64(
                18,
            ),
            Static(
                "thread.name",
            ): String(
                Owned(
                    "tokio-runtime-worker",
                ),
            ),
            Static(
                "queue_sizes",
            ): String(
                Owned(
                    "{DataId(\"op/latency\"): 10}",
                ),
            ),
            Static(
                "idle_ns",
            ): I64(
                244581990,
            ),
            Static(
                "code.filepath",
            ): String(
                Static(
                    "binaries/daemon/src/listener/tcp.rs",
                ),
            ),
            Static(
                "code.namespace",
            ): String(
                Static(
                    "dora_daemon::listener::tcp",
                ),
            ),
            Static(
                "busy_ns",
            ): I64(
                60210,
            ),
        },
        evict_list: [
            Static(
                "idle_ns",
            ),
            Static(
                "busy_ns",
            ),
            Static(
                "queue_sizes",
            ),
            Static(
                "thread.name",
            ),
            Static(
                "thread.id",
            ),
            Static(
                "code.lineno",
            ),
            Static(
                "code.namespace",
            ),
            Static(
                "code.filepath",
            ),
        ],
        max_len: 128,
        dropped_count: 0,
    },
    events: EvictedQueue {
        queue: None,
        max_len: 128,
        dropped_count: 0,
    },
    links: EvictedQueue {
        queue: None,
        max_len: 128,
        dropped_count: 0,
    },
    status: Unset,
    resource: Resource {
        attrs: {
            Static(
                "service.name",
            ): String(
                Owned(
                    "unknown_service",
                ),
            ),
        },
        schema_url: None,
    },
    instrumentation_lib: InstrumentationLibrary {
        name: "opentelemetry",
        version: Some(
            "0.18.0",
        ),
        schema_url: None,
    },
}
```

which is *better* but the coloring and the pretty format from tokio is not working. The log is quite verbose as well and might be hard to use as-it-is without Jaeger.

Comment (haixuanTao) [reply libraries/extensions/telemetry/tracing/src/lib.rs]: Absolutely. I think it is the most natural use case.

**Review by phil-opp — APPROVED** [review]
Thanks!

**Pull request merged**